src/mutation.rs

```diff
@@ -0,0 +1,881 @@
+    async fn place_order<'a>(
+        &self,
+        ctx: &Context<'a>,
+        #[graphql(desc = "Uuid of order to place")] id: Uuid,
+    ) -> Result<Order> {
+        let db_client = ctx.data_unchecked::<Database>();
+        let collection: Collection<Order> = db_client.collection::<Order>("orders");
+        let order = query_order(&collection, id).await?;
+        authenticate_user(&ctx, order.user._id)?;
+        set_status_placed(&collection, id).await?;
+        send_order_created_event(order).await?;
+        query_order(&collection, id).await
+    }
+}
+
+/// Calculates the total compensatable amount of all order items in the input by summing up their `compensatable_amount` attributes.
+fn calculate_compensatable_order_amount(order_items: &Vec<OrderItem>) -> u64 {
+    order_items.iter().map(|o| o.compensatable_amount).sum()
+}
+
+/// Extracts UUID from Bson.
+///
+/// Creating a order returns a UUID in a Bson document. This function helps to extract the UUID.
+fn uuid_from_bson(bson: Bson) -> Result<Uuid> {
+    match bson {
+        Bson::Binary(id) => Ok(id.to_uuid()?),
+        _ => {
+            let message = format!(
+                "Returned id: `{}` needs to be a Binary in order to be parsed as a Uuid",
+                bson
+            );
+            Err(Error::new(message))
+        }
+    }
+}
+
+/// Sets the status of an order to `OrderStatus::Placed`.
+/// Checks if pending order is still valid before setting `OrderStatus::Placed`.
+/// Rejects order if timestamp of placement exceeds `PENDING_TIMEOUT` in relation to the order creation timestamp.
+///
+/// * `collection` - MongoDB collection to update.
+/// * `input` - `UpdateOrderInput`.
+async fn set_status_placed(collection: &Collection<Order>, id: Uuid) -> Result<()> {
+    let current_timestamp_system_time = SystemTime::now();
+    let order = query_object(&collection, id).await?;
+    let order_created_at_system_time = order.created_at.to_system_time();
+    if order_created_at_system_time + PENDING_TIMEOUT >= current_timestamp_system_time {
+        let current_timestamp = DateTime::from(current_timestamp_system_time);
+        set_status_placed_in_mongodb(&collection, id, current_timestamp).await
+    } else {
+        set_status_rejected_in_mongodb(&collection, id).await
+    }
+}
+
+/// Updates order to `OrderStatus::Placed` in MongoDB.
+async fn set_status_placed_in_mongodb(
+    collection: &Collection<Order>,
+    id: Uuid,
+    current_timestamp: DateTime,
+) -> Result<()> {
+    let result = collection
+        .update_one(
+            doc! {"_id": id },
+            doc! {"$set": {"order_status": OrderStatus::Placed, "placed_at": current_timestamp}},
+            None,
+        )
+        .await;
+    if let Err(_) = result {
+        let message = format!("Placing order of id: `{}` failed in MongoDB.", id);
+        return Err(Error::new(message));
+    }
+    Ok(())
+}
+
+/// Updates order to `OrderStatus::Rejected` in MongoDB.
```

Comment:
What do you exactly mean by that?

Reply:
well the purpose of this is afaik to change the status in mongodb and then return an err which will be passed through
the docs only mention the first part, and imho if a function always returns an err, this could be mentioned